verifier/src/qe_report_body.rs

```diff
@@ -0,0 +1,548 @@
+        }
+    }
+}
+
+impl<E: Accessor<IsvSvn>> Verifier<E> for QeIsvSvnVerifier {
+    type Value = (IsvSvn, TcbLevel);
+    fn verify(&self, evidence: &E) -> VerificationOutput<Self::Value> {
+        let isv_svn = evidence.get();
+        let tcb_levels = &self.tcb_levels;
+        let tcb_level = tcb_levels
+            .iter()
+            .find_map(|tcb_level| {
+                if tcb_level.isv_svn().as_ref() <= isv_svn.as_ref() {
+                    Some(tcb_level.clone())
+                } else {
+                    None
+                }
+            })
+            .unwrap_or_default();
+
+        let is_success_bool =
+            tcb_level.status() == AdvisoryStatus::UpToDate && tcb_level.advisory_ids().is_empty();
+        let is_success = if is_success_bool { 1 } else { 0 };
+
+        VerificationOutput::new((isv_svn, tcb_level), is_success.into())
+    }
+}
+
+impl Display for QeIsvSvnVerifier {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        write!(
+            f,
+            "The {} should correspond to an `UpToDate` level with no advisories, from: {:?}",
+            IsvSvn::spaced_struct_name(),
+            self.tcb_levels
+        )
+    }
+}
+
+impl VerificationMessage<(IsvSvn, TcbLevel)> for QeIsvSvnVerifier {
+    fn fmt_padded(
+        &self,
+        f: &mut Formatter<'_>,
+        pad: usize,
+        output: &VerificationOutput<(IsvSvn, TcbLevel)>,
+    ) -> core::fmt::Result {
+        let is_success = output.is_success();
+        let status = choice_to_status_message(is_success);
+        write!(f, "{:pad$}{status} {self}", "")?;
+        if (!is_success).into() {
+            let name = IsvSvn::spaced_struct_name();
+            let actual = &output.value;
+            write!(
+                f,
+                ", but the {name} of {} corresponds to the {:?}",
+                actual.0, actual.1
+            )?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::qe_identity::{SignedQeIdentity, Tcb};
+    use crate::VerificationTreeDisplay;
+    use alloc::format;
+    use alloc::string::ToString;
+    use mc_sgx_core_sys_types::sgx_report_body_t;
+    use mc_sgx_core_types::AttributeFlags;
+    use mc_sgx_dcap_types::Quote3;
+
+    #[test]
+    fn no_tcb_levels_for_qe_isv_svn_fails() {
+        let isv_svn = IsvSvn::from(1);
+        let verifier = QeIsvSvnVerifier::new(&[]);
+        let verification = verifier.verify(&isv_svn);
+        assert_eq!(verification.is_failure().unwrap_u8(), 1);
+
+        let displayable = VerificationTreeDisplay::new(&verifier, verification);
+        assert_eq!(
+            displayable.to_string(),
+            r#"- [ ] The ISV SVN should correspond to an `UpToDate` level with no advisories, from: [], but the ISV SVN of 1 corresponds to the TcbLevel { tcb: Tcb { isv_svn: 0 }, tcb_date: "1970-01-01T00:00:00Z", tcb_status: OutOfDate, advisory_ids: [] }"#
```

Comment:
This message goes back to my comment early where I provided a default, over an Option<TcbLevel> it might be better to display `(none)` or similar instead of a default that the user may be wondering where it came from.

Reply:
Is this referring to the entire TcbLevel? e.g. `Option<TcbLevel>`?

Reply:
Yes, sorry there should have been more context in these comments

I've since moved over to the `Option<TcbLevel>`